### PR TITLE
Add some comments, add `can_define_opaque_ty` check to `try_normalize_ty_recur`

### PR DESCRIPTION
**Pull request by compiler-errors**
Follow-up from #117278, since I was recently re-reviewing this code.

### CODE REVIEW

**Comment by compiler-errors** [comment]
woops r? lcnr

File: compiler/rustc_trait_selection/src/solve/mod.rs [hunk 1]
```diff
@@ -313,19 +313,24 @@ impl<'tcx> EvalCtxt<'_, 'tcx> {
...
-                    return Some(ty);
-                }
+        // We do no always define opaque types eagerly to allow non-defining uses
+        // in the defining scope. However, if we can unify this opaque to an existing
+        // opaque, then we should attempt to eagerly reveal the opaque, and we fall
+        // through.
+        if let DefineOpaqueTypes::No = define_opaque_types
+            && let ty::Opaque = kind
+            && let Some(def_id) = alias.def_id.as_local()
+            && self.can_define_opaque_ty(def_id)
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/solve/mod.rs:323]: I guess this was previously incorrect for `Reveal::All`? :thinking:

File: compiler/rustc_trait_selection/src/solve/alias_relate.rs [hunk 2]
```diff
@@ -45,11 +40,18 @@ impl<'tcx> EvalCtxt<'_, 'tcx> {
                 self.evaluate_added_goals_and_make_canonical_response(Certainty::Yes)
             }
 
+            // 1. When we have an alias being related to an infer var, then assign
+            // the type (or const) of the alias to the infer var.
+            // 2. When we have an opaque being related to a rigid type (which, due to 1,
+            // is not an infer var), then assign the hidden type of the opaque to be
+            // the rigid type.
+            // 3. Otherwise, a rigid projection does not equal a concrete type ever.
```
Comment (lcnr) [inline compiler/rustc_trait_selection/src/solve/alias_relate.rs:48]: should this be moved to the top of the match?

**Review by lcnr — COMMENTED** [review]
realized:

we really should mention how `AliasRelate(<_ as Ambig>::Assoc, whatever)` ends up normalizing the alias to `?unconstrained` and it doesn't really matter that we equate it with `whatever`, as these constraints do not flow out to the caller in any way.

**Comment by lcnr** [comment]
r=me after nits

File: compiler/rustc_trait_selection/src/solve/mod.rs (hunk 1 above)
Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/solve/mod.rs:323]: No, this is purely an optimization. In `Reveal::All`, we still don't have any opaques in the opaque storage.

Comment (lcnr) [reply compiler/rustc_trait_selection/src/solve/mod.rs:323]: my thought is: if the storage is empty we don't normalize opaque types. So for reveal all we did not previously normalize opaque types here

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/solve/mod.rs:323]: Oh -- that still happens then? We don't look at the `Reveal` mode in `can_define_opaque_ty`.

Comment (compiler-errors) [reply compiler/rustc_trait_selection/src/solve/mod.rs:323]: We do use `DefiningAnchor::Error` in codegen, though.

**Comment by compiler-errors** [comment]
> we really should mention how AliasRelate(<_ as Ambig>::Assoc, whatever) ends up normalizing the alias to ?unconstrained and it doesn't really matter that we equate it with whatever, as these constraints do not flow out to the caller in any way.

Is the point you're trying to make that as long as we have an ambiguous alias in an alias-relate, we don't ever have inference side-effects and it'll always be ambiguous? In that case, sure, I agree, but I also don't know where we could best document this...

**Comment by compiler-errors** [comment]
TODO: I'll move all this to a module-level doc. I'll also add a comment to compute_projection_goal explaining why alias-relate is being emitted directly.

**Comment by compiler-errors** [comment]
For the record, I did rewrite the comment at the top of the alias-relate file, so I thought you wanted to read that. But thx for the r+ anyways.

**Pull request merged**